### ISSUE TITLE
Mac Compiling issue

### ISSUE THREAD

**Issue by AndersonChang** [openMVG/openMVG#1727]
Hi all,

I tried to follow the steps :

1.  git clone --recursive https://github.com/openMVG/openMVG.git
2. mkdir openMVG_Build
3. cd openMVG_Build
4. cmake -DCMAKE_BUILD_TYPE=RELEASE . ../openMVG/src/
5. make install 

and then I got the following message:

[  0%] Building HTML documentation with Sphinx
Running Sphinx v2.4.0
WARNING: html_static_path entry '_static' does not exist
loading pickled environment... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
build succeeded, 1 warning.

The HTML pages are in htmlDoc.
[  0%] Built target doc
[ 14%] Built target lib_CoinUtils
[ 17%] Built target lib_Osi
[ 30%] Built target lib_clp
[ 31%] Built target lib_OsiClpSolver
[ 32%] Built target openMVG_stlplus
[ 34%] Built target openMVG_lemon
[ 34%] Built target main_svgSample
[ 35%] Built target openMVG_easyexif
[ 37%] Built target openMVG_fast
[ 37%] Built target openMVG_exif
[ 39%] Built target openMVG_features
[ 40%] Built target openMVG_numeric
[ 41%] Built target openMVG_image
[ 42%] Built target openMVG_linearProgramming
[ 48%] Built target openMVG_multiview
[ 49%] Built target openMVG_lInftyComputerVision
[ 51%] Built target openMVG_geometry
[ 52%] Built target openMVG_matching
[ 53%] Built target openMVG_kvld
[ 54%] Built target openMVG_matching_image_collection
[ 54%] Built target openMVG_multiview_test_data
[ 54%] Built target openMVG_robust_estimation
[ 54%] Built target openMVG_system
[ 60%] Built target openMVG_sfm
[ 60%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_cameras_undistoBrown
ld: library not found for -lgomp
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Darwin-x86_64-RELEASE/openMVG_sample_cameras_undistoBrown] Error 1
make[1]: *** [openMVG_Samples/cameras_undisto_Brown/CMakeFiles/openMVG_sample_cameras_undistoBrown.dir/all] Error 2
make: *** [all] Error 2

It seems the linker cannot find the target, could anyone instruct me how to modify the CmakeList.txt to solve this problem?

Thanks.

**Comment by pmoulon** [comment]
Temporary solution:
- deactivate OpenMP by adding `-DOpenMVG_USE_OPENMP=OFF` to your make command line

Long term solution:
- Understand why OpenMP is detected but why the library is not found.
See here for the logic, https://github.com/openMVG/openMVG/blob/develop/src/CMakeLists.txt#L133
Perhaps omp will be found on your system but not gomp.

Recent CMake seems to have new ways to handle OpenMP but it was not stable on all the machine I tried.

**Comment by AndersonChang** [comment]
Thanks pmoulon, I tried to recompile with the following command,

cmake -DCMAKE_BUILD_TYPE=RELEASE -DOpenMVG_USE_OPENMP=OFF -DOpenMVG_BUILD_TESTS=ON -DOpenMVG_BUILD_EXAMPLES=ON . ../openMVG/src/

which disabled OpenMP, and run make install :

Scanning dependencies of target openMVG_test_progress
[0~80%] 
[ 80%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_image_matching
[ 80%] Built target openMVG_sample_image_matching
Scanning dependencies of target openMVG_sample_image_matching_gmsfilter
[ 80%] Building CXX object openMVG_Samples/features_image_matching_gmsfilter/CMakeFiles/openMVG_sample_image_matching_gmsfilter.dir/describe_and_match_gmsfilter.cpp.o
[ 80%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_image_matching_gmsfilter
[ 80%] Built target openMVG_sample_image_matching_gmsfilter
Scanning dependencies of target openMVG_sample_describe_and_match_GUI_autogen
[ 80%] Automatic MOC for target openMVG_sample_describe_and_match_GUI
[ 80%] Built target openMVG_sample_describe_and_match_GUI_autogen
Scanning dependencies of target openMVG_sample_describe_and_match_GUI
[ 80%] Building CXX object openMVG_Samples/describe_and_match_GUI/CMakeFiles/openMVG_sample_describe_and_match_GUI.dir/openMVG_sample_describe_and_match_GUI_autogen/mocs_compilation.cpp.o
[ 80%] Building CXX object openMVG_Samples/describe_and_match_GUI/CMakeFiles/openMVG_sample_describe_and_match_GUI.dir/main.cpp.o
[ 80%] Building CXX object openMVG_Samples/describe_and_match_GUI/CMakeFiles/openMVG_sample_describe_and_match_GUI.dir/ImageView.cpp.o
[ 81%] Building CXX object openMVG_Samples/describe_and_match_GUI/CMakeFiles/openMVG_sample_describe_and_match_GUI.dir/MainWindow.cpp.o
[ 81%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_describe_and_match_GUI.app/Contents/MacOS/openMVG_sample_describe_and_match_GUI
[ 81%] Built target openMVG_sample_describe_and_match_GUI
Scanning dependencies of target openMVG_sample_geodesy_show_exif_gps_position
[ 81%] Building CXX object openMVG_Samples/geodesy_show_exif_gps_position/CMakeFiles/openMVG_sample_geodesy_show_exif_gps_position.dir/show_exif_gps_position_demo.cpp.o
[ 81%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_geodesy_show_exif_gps_position
[ 81%] Built target openMVG_sample_geodesy_show_exif_gps_position
Scanning dependencies of target openMVG_sample_pano_converter
[ 81%] Building CXX object openMVG_Samples/image_spherical_to_pinholes/CMakeFiles/openMVG_sample_pano_converter.dir/main_pano_converter.cpp.o
[ 81%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_pano_converter
[ 81%] Built target openMVG_sample_pano_converter
Scanning dependencies of target openMVG_sample_image_undistort_autogen
[ 81%] Automatic MOC for target openMVG_sample_image_undistort
[ 81%] Built target openMVG_sample_image_undistort_autogen
Scanning dependencies of target openMVG_sample_image_undistort
[ 81%] Building CXX object openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/openMVG_sample_image_undistort_autogen/mocs_compilation.cpp.o
[ 81%] Building CXX object openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/main.cc.o
[ 81%] Building CXX object openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/MainWindow.cc.o
[ 82%] Building CXX object openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/QImageInterface.cc.o
[ 82%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_sample_image_undistort
Undefined symbols for architecture x86_64:
  "image_undistort_gui::MainWindow::staticMetaObject", referenced from:
      image_undistort_gui::MainWindow::onOpenImage() in MainWindow.cc.o
      image_undistort_gui::MainWindow::onExportImage() in MainWindow.cc.o
  "vtable for image_undistort_gui::MainWindow", referenced from:
      _main in main.cc.o
      image_undistort_gui::MainWindow::MainWindow(QWidget*) in MainWindow.cc.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Darwin-x86_64-RELEASE/openMVG_sample_image_undistort] Error 1
make[1]: *** [openMVG_Samples/image_undistort_gui/CMakeFiles/openMVG_sample_image_undistort.dir/all] Error 2
make: *** [all] Error 2


Somehow it keeps stuck in openMVG_sample_image_undistort linker error,  it seems to me 
that some linker setting in CmakeList cannot find the path correctly.

Let me know if you have any suggestion.

Thanks.

**Comment by pmoulon** [comment]
For this issue, you can see here for the putative fix https://github.com/openMVG/openMVG/issues/1720

**Comment by AndersonChang** [comment]
I see , after I add "-DOpenMVG_BUILD_GUI_SOFTWARES=OFF" to disable the gui, it works for compiling src code. Thanks a lot ! : )

**Comment by pmoulon** [comment]
Without OpenMP OpenMVG will be a bit slow, I would advise you to use OpenMVG in a docker to use OpenMVG at its full speed on your Mac

**Comment by rperrot** [comment]
An alternative to build with openMP is to install a clang compiler with openMP enabled. That's possible using macports (maybe homebrew) and it works fine with openMVG.

**Comment by virtualritz** [comment]
The Apple Clang compiler does support OpenMP. I have a similar issue on macOS 10.15.4. This is IMHO related to header files not being found, not related to OpenMP support in the compiler:
```
[ 37%] Building CXX object openMVG/features/CMakeFiles/openMVG_features.dir/akaze/AKAZE.cpp.o
In file included from /Users/moritz/code/openMVG/src/openMVG/features/akaze/AKAZE.cpp:9:
In file included from /Users/moritz/code/openMVG/src/openMVG/features/akaze/AKAZE.hpp:39:
In file included from /Users/moritz/code/openMVG/src/openMVG/image/image_container.hpp:12:
In file included from /usr/local/include/eigen3/Eigen/Dense:1:
/usr/local/include/eigen3/Eigen/Core:266:10: fatal error: 'omp.h' file not found
#include <omp.h>
```

**Comment by pmoulon** [comment]
@virtualritz You should access this line only if `EIGEN_HAS_OPENMP` is defined...
This preprocessor should not be defined by default from my understanding...

```
#if (defined _OPENMP) && (!defined EIGEN_DONT_PARALLELIZE)
  #define EIGEN_HAS_OPENMP
#endif
```

So you could fix your build by using -DEIGEN_DONT_PARALLELIZE=1 in the make command line of OpenMVG

**Comment by virtualritz** [comment]
I tried:
```terminal
> cmake -DCMAKE_BUILD_TYPE=RELEASE -DEIGEN_DONT_PARALLELIZE=1 ../src
...
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    EIGEN_DONT_PARALLELIZE


-- Build files have been written to: /Users/moritz/code/openMVG/build
> make -j4
...
[ 36%] Building CXX object openMVG/numeric/CMakeFiles/openMVG_numeric.dir/numeric.cpp.o
In file included from /Users/moritz/code/openMVG/src/openMVG/numeric/nullspace.cpp:9:
In file included from /Users/moritz/code/openMVG/src/openMVG/numeric/nullspace.hpp:12:
In file included from /Users/moritz/code/openMVG/src/openMVG/numeric/eigen_alias_definition.hpp:17:
In file included from /usr/local/include/eigen3/Eigen/Dense:1:
/usr/local/include/eigen3/Eigen/Core:266:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
```

**Comment by virtualritz** [comment]
I also tried out commenting our the ` #define EIGEN_HAS_OPENMP`. No luck. Eigen keeps looking for the `omp.h`.

**Comment by pmoulon** [comment]
My bad, to forwards the preprocessor to the compiler you need to use https://cmake.org/cmake/help/v3.0/command/add_definitions.html.

**Comment by virtualritz** [comment]
The bottom line is that the macOS build of OpenMVG is broken for users of Apple Clang.

I'm building [darktable](www.darktable.org) with Apple Clang with OpenMP support w/o issues on my machine. Aka: this is not a compiler issue.

**Comment by virtualritz** [comment]
> My bad, to forwards the preprocessor to the compiler you need to use https://cmake.org/cmake/help/v3.0/command/add_definitions.html.

Well, as I said: I commented out that line with the `#define` and it was still looking for that file. So this doesn't work. The only thing that does is to specify `-DOpenMVG_USE_OPENMP=OFF` but that makes OpenMVG unusably slow on my box.

**Comment by rperrot** [comment]
> The bottom line is that the macOS build of OpenMVG is broken for users of Apple Clang. Using OpenMVG w/o OpenMP is too slow to be feasible.

> Aka: this is not a compiler issue.

There are many issues :
1. Apple clang doesn't come with OpenMP support (maybe il will never support it) ;
2. Macports (and maybe homebrew) clang includes openMP support and it is fully supported on macOS.
3. No it's not a compiler issue, it's detection of openMP on a dependency of openMVG (Eigen) that cause this issue. 

>  The only thing that does is to specify -DOpenMVG_USE_OPENMP=OFF but that makes OpenMVG unusably slow on my box.

There is no alternative of using a different compiler (aka macports clang) if you want openMP support. If you want to keep apple clang you must agree with it's drawback.

**Comment by pmoulon** [comment]
Sidenote: You can use OpenMVG in a Docker and so still using all the core of your machine to run OpenMVG

**Comment by virtualritz** [comment]
> There is no alternative of using a different compiler (aka macports clang) if you want openMP support. If you want to keep apple clang you must agree with it's drawback.

You are mistaken. As I wrote above – I am building darktable (which needs OpenMP) on the same system using _Apple_ Clang. See e.g. [here](https://iscinumpy.gitlab.io/post/omp-on-high-sierra/) for how to do this. Again. This is _not_ a compiler issue.

TLDR; Try this on using Apple Clang:
```
clang -Xpreprocessor -fopenmp foo.c -lomp
```

**Comment by rperrot** [comment]
Thanks for your suggestion, but it's actually building openMP (using llvm version) to bring support to apple clang. That's exactly the same as macports/homebrew clang version did in a more cleaner way.

Since you compile an openMP version that is not exactly the same version as the LLVM backend you may encounter performance degradation (because of some LTO that may not be possible). Strong/deep integration of openMP is always preferable. 

Even it works, i strongly suggest that you use a clean install of llvm/clang/libomp from the same source/version for maximum performance and compatibility.

**Comment by virtualritz** [comment]
The issue is that there are tons of issues with building stuff when you have the Homebrew (or MacPorts clang) installed next to Apple clang. I've been fighting with them over two years and I do not want to go back there. I'm usually avoiding building stuff with OpenMP on macOS and just switch to Linux but at the moment I'm far from home and stuck with an old MBP that doesn't have the space to install even a VM.

**Comment by rperrot** [comment]
> I'm usually avoiding building stuff with OpenMP 

I agree. We try our best to avoid openMP in openMVG (using std::thread for ex), but it' a tedious task.

**Comment by cogitas3d** [comment]
Im with this problem on Mac OS X Catalina.

I'm using develop branch:
git clone --recursive https://github.com/openMVG/openMVG.git -b develop

I use the flags:
cmake -DCMAKE_BUILD_TYPE=RELEASE  -DOpenMVG_BUILD_GUI_SOFTWARES=OFF -DOpenMVG_USE_OPENMP=OFF ../src

But I had this error:

[ 80%] Built target openMVG_sample_pano_spherical_to_cubic
[ 81%] Building CXX object software/SfM/CMakeFiles/openMVG_main_SplitMatchFileIntoMatchFiles.dir/main_SplitMatchFileIntoMatchFiles.cpp.o
[ 81%] Linking CXX executable ../../Darwin-x86_64-RELEASE/openMVG_main_SplitMatchFileIntoMatchFiles
[ 81%] Built target openMVG_main_SplitMatchFileIntoMatchFiles
[ 82%] Building CXX object software/SfM/CMakeFiles/openMVG_main_ComputeStructureFromKnownPoses.dir/main_ComputeStructureFromKnownPoses.cpp.o
/Users/ciceromoraes/Programs/openMVG/src/software/SfM/main_ComputeStructureFromKnownPoses.cpp:33:10: fatal error: 
      'ceres/types.h' file not found
#include <ceres/types.h>
         ^~~~~~~~~~~~~~~

This is the first time in years, that I have problem with openMVG compilation.

**Comment by cogitas3d** [comment]
[SOLVED]

How I need only some of the coomand, i solved the problem ignoring the make errors:

**make -i**

Thank you all to show how to solve the other problems!

**Comment by pmoulon** [comment]
@cogitas3d Normally the problem should be solved with this https://github.com/openMVG/openMVG/blob/develop/src/software/SfM/CMakeLists.txt#L154

Can you add `message(${CERES_INCLUDE_DIRS})` in the cmakelist and see if it is having a value?

**Comment by rfabbri** [comment]
@virtualritz please see PR #1908 for a fix on the `develop` branch for the include problem `'omp.h' file not found`
The develop branch with this fix has no link problems on Mac OS using Xcode 13.3 under
macOS Monterey 12.2.1 (march/2022).
